plugins/woocommerce/src/Admin/API/OnboardingPlugins.php

```diff
@@ -183,6 +195,38 @@ function( $action ) use ( $job_id ) {
 		return $response;
 	}
 
+
+	/**
+	 * Return Jetpack authorization URL.
+	 *
+	 * @return array
+	 * @throws \Exception
+	 */
+	public function get_jetpack_authorization_url() {
+		$manager = new Manager( 'woocommerce' );
+		// Register the site to wp.com.
+		if ( ! $manager->is_connected() ) {
+
+			$result = $manager->try_registration();
+			if ( is_wp_error( $result ) ) {
+				 throw new \Exception( $result->get_error_message() );
+			}
+		}
+
+		$redirect_url = apply_filters( 'woocommerce_admin_onboarding_jetpack_connect_redirect_url', esc_url_raw( admin_url( 'admin.php?page=wc-admin' ) ) );
+		$calypso_env  = defined( 'WOOCOMMERCE_CALYPSO_ENVIRONMENT' ) && in_array( WOOCOMMERCE_CALYPSO_ENVIRONMENT, [ 'development', 'wpcalypso', 'horizon', 'stage' ], true ) ? WOOCOMMERCE_CALYPSO_ENVIRONMENT : 'production';
+
+		return [
+			'url' => add_query_arg(
+				[
+					'from'        => apply_filters( 'woocommerce_admin_onboarding_jetpack_connect_from_arg', 'woocommerce-onboarding' ),
+					'calypso_env' => $calypso_env,
+				],
+				$manager->get_authorization_url( null, $redirect_url )
+			),
+		];
```

Comment:
Can the `redirect_url` and `from` values be passed to the endpoint as parameters?

Looking into the future a bit, I think the connection screen will need to render differently (progress bar, copy, images) depending on what the user was doing:

- Selecting Jetpack in Core Profiler
- Setting up mobile app
- Using an AI feature
- Setting up shipping / mobile task

I'm thinking that each of those features above would perform the REST API request with granular `redirect_url` and `from` values relating to the feature. If we did that, the connection screen could then be customized for each feature as needed, and have redirection back to where the user started from.

I think maybe you had the same idea with the filters but I'm not sure it can support this level of granularity?

Reply:
I agree, I think it might be better as URL parameters. Using filters requires PHP for it to work, and I think it risks "leaks" (i.e.: A plugin unintentionally registers the filter in a global code), whereas URL param requires more deliberate use.

Reply:
Thank you both! Updated in https://github.com/woocommerce/woocommerce/pull/38674/commits/a5f3581dcd6772d72386593a745397779fdef087